### PR TITLE
NF-457 - validate address post lookup

### PR DESCRIPTION
**Pull request by SteveSugden**

### CODE REVIEW

File: app/uk/gov/hmrc/nationalimportdutyadjustmentcentrefrontend/controllers/makeclaim/ImporterDetailsController.scala [hunk 1]
```diff
@@ -95,11 +95,18 @@ class ImporterDetailsController @Inject() (
...
           ImporterContactDetails(el._1, el._2, el._3, el._4, confirmedAddress.address.postcode.getOrElse(""))
-        val auditableImporterContactDetails = AuditableImporterContactDetails(updatedAddress, confirmedAddress.auditRef)
-        data.updateCreateAnswers(
-          answers => answers.copy(importerContactDetails = Some(auditableImporterContactDetails))
-        ) map {
-          _ => Redirect(nextPage(answers))
+        // Address Lookup Service may return an address that is incompatible with NIDAC, so validate it again
+        val formWithAddress = form.fillAndValidate(updatedAddress)
+        if (formWithAddress.hasErrors)
+          Future.successful(BadRequest(detailsView(formWithAddress, backLink(answers))))
```
Comment (njmoorewitchcat) [inline app/uk/gov/hmrc/nationalimportdutyadjustmentcentrefrontend/controllers/makeclaim/ImporterDetailsController.scala:101]: AMBER:  Should we save the address and redirect to the edit page?  This current implementation leaves the user on this url:  /apply-for-return-import-duty-paid-on-deposit-or-guarantee/create/your-address/update?id=b91448d0-7dcd-478b-849f-3bf0ba6313f5  With all the update gubbins on the url (when the address fails).

Comment (njmoorewitchcat) [reply app/uk/gov/hmrc/nationalimportdutyadjustmentcentrefrontend/controllers/makeclaim/ImporterDetailsController.scala:101]: It also means it becomes part of the browser 'back' history

File: test/uk/gov/hmrc/nationalimportdutyadjustmentcentrefrontend/controllers/makeclaim/AddressControllerSpec.scala [hunk 2]
```diff
@@ -122,27 +121,41 @@ class AddressControllerSpec extends ControllerSpec with TestData {
...
+    "preserve audit ref and redirect when same answer is submitted" in {
+
+      val previouslyLookedUpAddress = auditableAddress
+
+      withCacheCreateAnswers(emptyAnswers.copy(claimantAddress = Some(previouslyLookedUpAddress)))
+
+      val result = controller.onSubmit()(validRequest)
+      status(result) mustEqual SEE_OTHER
+      theUpdatedCreateAnswers.claimantAddress mustBe Some(auditableAddress)
+      theUpdatedCreateAnswers.claimantAddress.flatMap(ca => ca.auditRef) mustBe auditableAddress.auditRef
```
Comment (rorticus0502) [inline test/uk/gov/hmrc/nationalimportdutyadjustmentcentrefrontend/controllers/makeclaim/AddressControllerSpec.scala:155]: if the line above is satisfied then this line will be satisfied

File: test/uk/gov/hmrc/nationalimportdutyadjustmentcentrefrontend/controllers/makeclaim/ImporterDetailsControllerSpec.scala [hunk 3]
```diff
@@ -143,6 +150,21 @@ class ImporterDetailsControllerSpec extends ControllerSpec with TestData {
...
+    "preserve audit ref and redirect when same answer is submitted" in {
+
+      val previouslyLookedUpAddress = auditableImporterContactDetails
+
+      withCacheCreateAnswers(emptyAnswers.copy(importerContactDetails = Some(previouslyLookedUpAddress)))
+
+      val result = controller.onSubmit()(validRequest)
+      status(result) mustEqual SEE_OTHER
+      theUpdatedCreateAnswers.importerContactDetails mustBe Some(auditableImporterContactDetails)
+      theUpdatedCreateAnswers.importerContactDetails.flatMap(
```
Comment (rorticus0502) [inline test/uk/gov/hmrc/nationalimportdutyadjustmentcentrefrontend/controllers/makeclaim/ImporterDetailsControllerSpec.scala:162]: correct me if I am wrong. I think this assertion is covered by above assertion

File: test/uk/gov/hmrc/nationalimportdutyadjustmentcentrefrontend/controllers/makeclaim/AddressControllerSpec.scala (hunk 2 above)
Comment (SteveSugden) [reply test/uk/gov/hmrc/nationalimportdutyadjustmentcentrefrontend/controllers/makeclaim/AddressControllerSpec.scala:155]: True.  I suppose this just emphasises the intent of the test - the audit ref is preserved.

File: test/uk/gov/hmrc/nationalimportdutyadjustmentcentrefrontend/controllers/makeclaim/ImporterDetailsControllerSpec.scala (hunk 3 above)
Comment (SteveSugden) [reply test/uk/gov/hmrc/nationalimportdutyadjustmentcentrefrontend/controllers/makeclaim/ImporterDetailsControllerSpec.scala:162]: You are not wrong!

Comment (rorticus0502) [reply test/uk/gov/hmrc/nationalimportdutyadjustmentcentrefrontend/controllers/makeclaim/ImporterDetailsControllerSpec.scala:162]: Yay

File: test/uk/gov/hmrc/nationalimportdutyadjustmentcentrefrontend/controllers/makeclaim/AddressControllerSpec.scala (hunk 2 above)
Comment (rorticus0502) [reply test/uk/gov/hmrc/nationalimportdutyadjustmentcentrefrontend/controllers/makeclaim/AddressControllerSpec.scala:155]: I thought the same thing. It does no harm.

**Review by rorticus0502 — APPROVED** [review]

**Pull request merged**